### DISCUSSION TITLE
MSSQL: DISTINCT + TOP syntax seems incorrect

### DISCUSSION THREAD

**Discussion by tboddyspargo** [sqlalchemy/sqlalchemy#11518] [Usage Questions] [answered]
It seems that the `mssql` (pyodbc) dialect may have the incorrect order for `DISTINCT` when `TOP` is also present. It's possible that other syntactic elements (beside just `TOP` may also be affected).

NOTE: I'm on sqlalchemy 1.4.46

### Actual
```
SELECT TOP 10 DISTINCT [x] FROM [test_table]]
```

### Expected

```
SELECT DISTINCT TOP 10 [x] FROM [test_table]]
```

### Test Script

```python
from sqlalchemy import Column, MetaData, Table, create_engine, distinct, select

distinct_query = select([distinct(Column("x"))]).select_from(Table("test_table", MetaData())).limit(10)


engine = create_engine(
    "mssql+pyodbc://sa:***@localhost:1433/test_db?driver=ODBC+Driver+18+for+SQL+Server&Encrypt=no&TrustServerCertificate=yes",
    echo=True,
)
engine.execute(distinct_query).fetchall()
```

```
Traceback (most recent call last):
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
pyodbc.ProgrammingError: ('42000', "[42000] [Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Incorrect syntax near the keyword 'DISTINCT'. (156) (SQLExecDirectW)")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/tyler/dev/datachat/pipeliner/tmp/distinct_issue.py", line 10, in <module>
    engine.execute(distinct_query).fetchall()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 2, in execute
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/util/deprecations.py", line 468, in warned
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 3257, in execute
    return connection.execute(statement, *multiparams, **params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1572, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/Users/tyler/dev/datachat/pipeliner/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', "[42000] [Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Incorrect syntax near the keyword 'DISTINCT'. (156) (SQLExecDirectW)")
[SQL: SELECT TOP 10 DISTINCT [x] 
FROM [test_table]]
(Background on this error at: https://sqlalche.me/e/14/f405)
```

I'm still looking into how I might be able to patch the dialect/compiler to work around this, but I thought I'd open the discussion to get some input.

**Comment by zzzeek** [comment, marked as answer]
this is the wrong use:

     distinct_query = select([distinct(Column("x"))]).select_from(Table("test_table", MetaData())).limit(0)

see note at https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.distinct:

> The [distinct()](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.distinct) function does not apply DISTINCT to the full SELECT statement, instead applying a DISTINCT modifier to individual column expressions. For general SELECT DISTINCT support, use the [Select.distinct()](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.distinct) method on [Select](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select).

use distinct()

     distinct_query = select([Column("x")]).distinct().select_from(Table("test_table", MetaData())).limit(0)

**Reply by tboddyspargo** [reply]
Thanks for your support, @zzzeek, as always! I'll explore how we are getting this wrong!

**Reply by zzzeek** [reply]
well it's on us because everyone makes this mistake.   maybe it's the doc search, not sure

**Reply by CaselIT** [reply]
would it be practical to warn if select sees in its clause an `UnaryExpression` with `distinct_op` as operator?

**Reply by zzzeek** [reply]
EDIT: mike is old, ignore   

how is that not the normal use of `distinct()`.   i think the idea is that this is a PostgreSQL construct only and maybe we establish that and literally move it out of sqlalchemy.sql and into dialects.postgresql.  so the main one can just emit deprecation warning.  hows that?

**Reply by zzzeek** [reply]
EDIT: mike is old, ignore   

then later standalone distinct() is just a CompileError on every other backend.

we have to establish that this is in fact only a PG thing though.

**Reply by CaselIT** [reply]
Isn't `function(distinct some_column)` sql standard?

Sqlite does support it https://www.sqlite.org/lang_aggfunc.html

Have not checked other dbs.

Note that this isn't the `distinct on` that pg only

**Reply by zzzeek** [reply]
OK !  ignore everything I just said, I was confusing this with `DISTINCT ON`.

This is distinct inside a function.  OK.  yes, let's make this warn/raise when it's used like this

**Reply by CaselIT** [reply]
how would we go about it?

should we do a subclass of UnaryExpression? or change the coercion logic to look for unary + distinct op?

**Reply by zzzeek** [reply]
you could do it at compile time by passing a kw flag that "distinct" looks for , indicating the appropriate context.   if flag not there, boom.   would have to find a way to make it OK for str() compile

the other way is yes do it at coercion time.   id have to look to see what route there is.

**Reply by CaselIT** [reply]
I think trying to do this at coercion time is likely better, but I is would be missed when not a top-level element. Meaning
```py
# would warn
select(distinct(column("x")))
# would not since the top-level is not the distinct
select(distinct(column("x")) + 1)

```

**Reply by zzzeek** [reply]
does "distinct()" only go inside of a SQL function ?   That's all we'd have to look for.  we can put "distinct()" into a special coercion category based on the roles, like `FunctionDistinctRole`.

**Reply by CaselIT** [reply]
Maybe. I'll open an issue so we can track it

**Reply by CaselIT** [reply]
follow on at #11526
